### PR TITLE
feat(topology): add poolTopologyAffinity/poolTopologySpread in Storage Class

### PR DESCRIPTION
**Pull request by sinhaashish**
Test case 

**When the poolTopologyAffinity has both key and value 
Then replica will be placed on pools matching both key and value** 
poolTopologyAffinity: "performance: Gold"
replica1: "performance: Gold"
replica2: "performance: Gold"

**When the poolTopologyAffinity has only key 
 Then replica will be placed on pools matching only the key , value can be anything**
poolTopologyAffinity: "performance"
replica1: "performance: Gold"
replica2: "performance: Silver"

**When the poolTopologySpread has both key and value
Then replica will be placed on pools matching only the key and not value**
poolTopologySpread: "storage: Hot". 
replica1: "storage: Cold"
replica2: "storage: Warm"

**When the poolTopologySpread has only key
Then replica will be placed on pools which does not match the key**
poolTopologySpread: "storage". 
replica1: "performance: Gold"
replica2: "performance: Silver"

| No: | Details   | Expected result   |
| :---: | :---: | :---: |
| 1|  the volume replicas has topology inclusion labels matching to the labels as that of certain pools   | volume creation should succeed with a returned volume object with topology   |
| 2|  the volume replicas has topology inclusion labels not matching with any pools with labels | volume creation should fail with an insufficient storage error   |
| 3|  the volume replicas has topology exclusion labels key matching with any pools with labels key and not value   | volume creation should succeed with a returned volume object with exclusion topology key same as pool key and different value   |
| 4|  the volume replicas has topology inclusion labels and exclusion labels same | replica creation should fail with an insufficient storage error   |

### CODE REVIEW

File: control-plane/agents/src/bin/core/controller/scheduling/volume_policy/pool.rs [hunk 1]
```diff
@@ -52,37 +52,77 @@ impl PoolBaseFilters {
...
+                            volume_pool_topology_exclusion_labels = labelled_topology.exclusion;
                         } else {
                             return true;
                         }
                     }
                 },
             },
         };
+
+        tracing::debug!(
```
Comment (tiagolobocastro) [inline control-plane/agents/src/bin/core/controller/scheduling/volume_policy/pool.rs]: Why do we need these logs?

File: control-plane/agents/src/bin/core/volume/operations_helper.rs [hunk 2]
```diff
@@ -490,7 +490,6 @@ impl OperationGuardArc<VolumeSpec> {
     ) -> Result<Vec<ReplicaId>, SvcError> {
         let mut created_replicas = Vec::with_capacity(count);
         let mut candidate_error = None;
-
```
Comment (tiagolobocastro) [inline control-plane/agents/src/bin/core/volume/operations_helper.rs]: unintended change?

File: control-plane/csi-driver/src/context.rs [hunk 3]
```diff
@@ -52,6 +52,10 @@ pub enum Parameters {
     CloneFsIdAsVolumeId,
     #[strum(serialize = "fsId")]
     FsId,
+    #[strum(serialize = "poolTopologyAffinity")]
```
Comment (tiagolobocastro) [inline control-plane/csi-driver/src/context.rs:56]: Do we need topology in the name?
ex: poolAffinity vs poolTopologyAffinity

File: control-plane/csi-driver/src/context.rs [hunk 4]
```diff
@@ -62,6 +66,26 @@ impl Parameters {
...
         })
     }
+    fn parse_map(
+        value: Option<&String>,
+    ) -> Result<Option<HashMap<String, String>>, serde_json::Error> {
+        Ok(match value {
+            Some(json_str) => {
+                let value: Value = serde_json::from_str(json_str)?;
+                // Convert Value to Map
+                if let Value::Object(map) = value {
```
Comment (tiagolobocastro) [inline control-plane/csi-driver/src/context.rs]:
```suggestion
                if let Some(obj) = value.as_object() {
```

Comment (tiagolobocastro) [reply control-plane/csi-driver/src/context.rs]: We should handle v.as_str().unwrap() better, should return an error rather than crash...

File: control-plane/csi-driver/src/context.rs [hunk 5]
```diff
@@ -173,6 +219,10 @@ impl PublishParams {
             publish_context.insert(Parameters::FsId.to_string(), fs_id.to_string());
         }
 
+        // if let Some(pool_topology_affinity) = self.pool_topology_affinity() {
+        //     publish_context.insert(Parameters::PoolTopologyAffinity.to_string(),
+        // pool_topology_affinity.to_string()); }
```
Comment (tiagolobocastro) [inline control-plane/csi-driver/src/context.rs]: is this meant to be not commented out?

File: k8s/operators/src/lib.rs [hunk 6]
```diff
@@ -2,6 +2,6 @@
 pub mod diskpool {
     /// The DiskPool custom resource definition.
     pub mod crd {
-        include!("pool/diskpool/v1beta1.rs");
+        include!("pool/diskpool/crd/v1beta2.rs");
```
Comment (tiagolobocastro) [inline k8s/operators/src/lib.rs:5]: this doesn't seem related? If not please move to separate commit or PR

**Review by tiagolobocastro — COMMENTED** [review]
Please separate into multiple commits (core-agent scheduling, csi-driver, etc)
Also tests are missing, please add them

File: control-plane/agents/src/bin/core/controller/scheduling/volume_policy/pool.rs [hunk 7]
```diff
@@ -52,37 +52,71 @@ impl PoolBaseFilters {
...
+            "volume_pool_topology_inclusion_labels {:?}",
+            volume_pool_topology_inclusion_labels
+        );
+        tracing::debug!(
+            "volume_pool_topology_exclusion_labels {:?}",
+            volume_pool_topology_exclusion_labels
+        );
```
Comment (Abhinandan-Purkait) [inline control-plane/agents/src/bin/core/controller/scheduling/volume_policy/pool.rs]: Don't think we need to put logs in scheduling methods?

File: control-plane/agents/src/bin/core/controller/scheduling/volume_policy/pool.rs [hunk 8]
```diff
@@ -52,37 +52,71 @@ impl PoolBaseFilters {
...
-                            .get(vol_key)
-                            .is_some_and(|pool_value| vol_val.is_empty() || pool_value == vol_val)
-                    }),
+                Some(label) => {
+                    // Return false if inclusion and exclusion labels are same.
+                    if volume_pool_topology_inclusion_labels
+                        == volume_pool_topology_exclusion_labels
+                    {
+                        return false;
+                    }
```
Comment (Abhinandan-Purkait) [inline control-plane/agents/src/bin/core/controller/scheduling/volume_policy/pool.rs]: This check should be move up, in line 71, i.e don't even need to proceed further from there?

File: control-plane/agents/src/bin/core/controller/scheduling/volume_policy/pool.rs [hunk 9]
```diff
@@ -52,37 +52,71 @@ impl PoolBaseFilters {
...
+
+                    volume_pool_topology_inclusion_labels.iter().all(
+                        |(vol_inc_key, vol_inc_val)| {
+                            // See `InclusiveLabel` doc comment.
+                            label.get(vol_inc_key).is_some_and(|pool_value| {
```
Comment (Abhinandan-Purkait) [inline control-plane/agents/src/bin/core/controller/scheduling/volume_policy/pool.rs]: nit: pool_value --> pool_inc_value, vol_inc_val --> vol_inc_value ?

File: control-plane/csi-driver/src/bin/controller/controller.rs [hunk 10]
```diff
@@ -260,9 +260,23 @@ impl rpc::csi::controller_server::Controller for CsiControllerSvc {
...
-        let mut inclusive_label_topology: HashMap<String, String> = HashMap::new();
+        // labels for pool inclusion
+        let mut pool_inclusive_label_topology: HashMap<String, String> = HashMap::new();
+        pool_inclusive_label_topology
+            .insert(String::from(CREATED_BY_KEY), String::from(DSP_OPERATOR));
+        if let Some(pool_topology_affinity) = context.publish_params().pool_topology_affinity() {
+            for (key, value) in pool_topology_affinity.iter() {
+                pool_inclusive_label_topology.insert(key.to_string(), value.to_string());
+            }
+        }
```
Comment (Abhinandan-Purkait) [inline control-plane/csi-driver/src/bin/controller/controller.rs]:
```suggestion
       // labels for pool inclusion
        let mut pool_inclusive_label_topology: HashMap<String, String> = HashMap::new();
        pool_inclusive_label_topology
            .insert(String::from(CREATED_BY_KEY), String::from(DSP_OPERATOR));
        pool_inclusive_label_topology.extend(
            context
                .publish_params()
                .pool_topology_affinity()
                .clone()
                .unwrap_or_default(),
        );
```

File: control-plane/csi-driver/src/bin/controller/controller.rs [hunk 11]
```diff
@@ -260,9 +260,23 @@ impl rpc::csi::controller_server::Controller for CsiControllerSvc {
...
 
-        inclusive_label_topology.insert(String::from(CREATED_BY_KEY), String::from(DSP_OPERATOR));
+        // labels for pool exclusion
+        let mut pool_exclusive_label_topology: HashMap<String, String> = HashMap::new();
+        if let Some(pool_topology_spread) = context.publish_params().pool_topology_spread() {
+            for (key, value) in pool_topology_spread.iter() {
+                pool_exclusive_label_topology.insert(key.to_string(), value.to_string());
+            }
+        }
```
Comment (Abhinandan-Purkait) [inline control-plane/csi-driver/src/bin/controller/controller.rs]: Not needed we can simply pass it ? As below

File: control-plane/csi-driver/src/bin/controller/controller.rs [hunk 12]
```diff
@@ -295,8 +309,8 @@ impl rpc::csi::controller_server::Controller for CsiControllerSvc {
                 let volume_topology = CreateVolumeTopology::new(
                     None,
                     Some(PoolTopology::labelled(LabelledTopology {
-                        exclusion: Default::default(),
-                        inclusion: inclusive_label_topology,
+                        exclusion: pool_exclusive_label_topology,
+                        inclusion: pool_inclusive_label_topology,
```
Comment (Abhinandan-Purkait) [inline control-plane/csi-driver/src/bin/controller/controller.rs]:
```suggestion
                        exclusion: context
                            .publish_params()
                            .pool_topology_spread()
                            .clone()
                            .unwrap_or_default(),
                        inclusion: pool_inclusive_label_topology,
```

File: control-plane/csi-driver/src/context.rs [hunk 13]
```diff
@@ -1,16 +1,16 @@
...
     collections::HashMap,
     convert::AsRef,
     num::ParseIntError,
     str::{FromStr, ParseBoolError},
 };
+use stor_port::types::v0::openapi::models::VolumeShareProtocol;
 use strum_macros::{AsRefStr, Display, EnumString};
-use tracing::log::warn;
+use tracing::warn;
+use utils::K8S_STS_PVC_NAMING_REGEX;
```
Comment (Abhinandan-Purkait) [inline control-plane/csi-driver/src/context.rs:13]: Please arrange the imports move line 10 and line 13 up with 1

File: control-plane/csi-driver/src/context.rs (hunk 3 above)
Comment (sinhaashish) [reply control-plane/csi-driver/src/context.rs:56]: It makes things look consistent like `poolTopologyAffinity` , `poolTopologySpread`, `nodeTopologyAffinity`, `nodeTopologySpread`

**Comment by sinhaashish** [comment]
> Please separate into multiple commits (core-agent scheduling, csi-driver, etc) Also tests are missing, please add them

added

File: control-plane/csi-driver/src/context.rs (hunk 3 above)
Comment (tiagolobocastro) [reply control-plane/csi-driver/src/context.rs:56]: Yes, good point

File: control-plane/agents/src/bin/core/controller/scheduling/volume_policy/pool.rs [hunk 14]
```diff
@@ -73,40 +73,88 @@ impl PoolBaseFilters {
...
-                    }),
-            },
+            Ok(spec) => {
+                match spec.labels {
+                    None => false,
+                    Some(label) => {
+                        // The inclusion labes of volume must match the labels in pool
+                        // for the specified pool to qualify for volume provisioning.
+                        // i.e both the maps volume_pool_topology_inclusion_labels and label must be
+                        // equal
```
Comment (tiagolobocastro) [inline control-plane/agents/src/bin/core/controller/scheduling/volume_policy/pool.rs]: This is not entirely true, it's possible for the pool to have more labels as long as doesn't conflict with the volume labels

File: control-plane/agents/src/bin/core/controller/scheduling/volume_policy/pool.rs [hunk 15]
```diff
@@ -73,40 +73,88 @@ impl PoolBaseFilters {
...
+                        let inclusion_match = label.iter().all(|(key, value)| {
+                            volume_pool_topology_inclusion_labels.get(key) == Some(value)
+                        }) && volume_pool_topology_inclusion_labels
+                            .iter()
+                            .all(|(key, value)| label.get(key) == Some(value));
```
Comment (tiagolobocastro) [inline control-plane/agents/src/bin/core/controller/scheduling/volume_policy/pool.rs]: Why is this required to match?

```suggestion
                        });
```

File: control-plane/agents/src/bin/core/controller/scheduling/volume_policy/pool.rs [hunk 16]
```diff
@@ -73,40 +73,88 @@ impl PoolBaseFilters {
...
+
+                        // The exclusion labels value in volume should not match the pool
+                        // labels value
+                        let common_keys: Vec<_> = volume_pool_topology_exclusion_labels
+                            .keys()
+                            .filter(|&key| label.contains_key(key))
```
Comment (tiagolobocastro) [inline control-plane/agents/src/bin/core/controller/scheduling/volume_policy/pool.rs]: what happens when the pool doesn't contain the exclusion label?

File: control-plane/agents/src/bin/core/controller/scheduling/volume_policy/pool.rs [hunk 17]
```diff
@@ -73,40 +73,88 @@ impl PoolBaseFilters {
...
+                            (true, false) => {
+                                return exclusion_match;
+                            }
+                            _ => {
+                                return true;
+                            }
+                        }
+                    }
+                }
+            }
```
Comment (tiagolobocastro) [inline control-plane/agents/src/bin/core/controller/scheduling/volume_policy/pool.rs:145]: can we split the bits above this into two functions, one for inclusion and one for exclusion?

File: control-plane/agents/src/bin/core/controller/scheduling/volume_policy/pool.rs [hunk 18]
```diff
@@ -73,40 +73,88 @@ impl PoolBaseFilters {
...
+                        volume_pool_topology_inclusion_labels.remove("openebs.io/created-by");
+
+                        match (
+                            volume_pool_topology_inclusion_labels.is_empty(),
+                            volume_pool_topology_exclusion_labels.is_empty(),
+                        ) {
+                            (false, false) => {
+                                return inclusion_match && exclusion_match;
+                            }
+                            (false, true) => {
```
Comment (tiagolobocastro) [inline control-plane/agents/src/bin/core/controller/scheduling/volume_policy/pool.rs]: can we return early by checking if the vec is empty before we start doing the exclusion calculation stuff?

File: control-plane/csi-driver/src/context.rs [hunk 19]
```diff
@@ -62,6 +67,26 @@ impl Parameters {
...
+        value: Option<&String>,
+    ) -> Result<Option<HashMap<String, String>>, serde_json::Error> {
+        Ok(match value {
+            Some(json_str) => {
+                let value: Value = serde_json::from_str(json_str)?;
+                // Convert Value to Map
+                if let Value::Object(map) = value {
+                    let map: HashMap<String, String> = map
+                        .into_iter()
+                        .map(|(k, v)| (k, v.as_str().unwrap_or_default().to_string()))
```
Comment (tiagolobocastro) [inline control-plane/csi-driver/src/context.rs]: Would this work?

```suggestion
                        .map(|(k, v)| (k, v.to_string()))
```

File: control-plane/csi-driver/src/context.rs [hunk 20]
```diff
@@ -62,6 +67,26 @@ impl Parameters {
...
+                        .collect();
+                    Some(map)
+                } else {
+                    None
+                }
```
Comment (tiagolobocastro) [inline control-plane/csi-driver/src/context.rs]: you can just map here, example: 

```suggestion
                value.as_object().map(|map| {
                    map
                        .into_iter()
                        .map(|(k, v)| (k, v.as_str().unwrap_or_default().to_string()))
                        .collect()
                })
```

File: control-plane/csi-driver/src/context.rs [hunk 21]
```diff
@@ -200,20 +247,32 @@ impl TryFrom<&HashMap<String, String>> for PublishParams {
                 .map_err(|_| tonic::Status::invalid_argument("Invalid keep_alive_tmo"))?;
         let fs_id = Parameters::fs_id(args.get(Parameters::FsId.as_ref()))
             .map_err(|_| tonic::Status::invalid_argument("Invalid fs_id"))?;
-
+        let pool_topology_affinity =
+            Parameters::pool_topology_affinity(args.get(Parameters::PoolTopologyAffinity.as_ref()))
+                .map_err(|_| {
+                    tonic::Status::invalid_argument("Invalid pool_topology_affinity timeout")
```
Comment (tiagolobocastro) [inline control-plane/csi-driver/src/context.rs]:
```suggestion
                    tonic::Status::invalid_argument("Invalid pool_topology_affinity")
```

File: control-plane/csi-driver/src/context.rs [hunk 22]
```diff
@@ -200,20 +247,32 @@ impl TryFrom<&HashMap<String, String>> for PublishParams {
...
+                })?;
+        let pool_topology_spread =
+            Parameters::pool_topology_spread(args.get(Parameters::PoolTopologySpread.as_ref()))
+                .map_err(|_| {
+                    tonic::Status::invalid_argument("Invalid pool_topology_spread timeout")
```
Comment (tiagolobocastro) [inline control-plane/csi-driver/src/context.rs]:
```suggestion
                    tonic::Status::invalid_argument("Invalid pool_topology_spread")
```

File: tests/bdd/features/volume/topology/pool-affinity.feature [hunk 23]
```diff
@@ -0,0 +1,30 @@
+Feature: Volume Topology
+  The volume topology feature is for selection of particular pools to schedule a replica on.
+  All the volume topology key and values should have an exact match with pool labels, although pools
```
Comment (tiagolobocastro) [inline tests/bdd/features/volume/topology/pool-affinity.feature]: If pools can have extra labels then it's not an "exact match".

File: tests/bdd/features/volume/topology/pool-affinity.feature [hunk 24]
```diff
@@ -0,0 +1,30 @@
...
+  can have extra labels, for the selection of that pool. No volume topology symbolizes the volume replicas
+  can be scheduled on any available pool.
+
+  Background:
+    Given a control plane, three Io-Engine instances, three pools
+
+  Scenario: suitable pools which contain volume topology labels
+    Given a request for a volume with its topology inclusion labels matching certain pools with same labels
```
Comment (tiagolobocastro) [inline tests/bdd/features/volume/topology/pool-affinity.feature]: I wonder if these might be too generic, maybe better with actual examples, and you can use Scenario Outline and Examples table for both input and outcome, wdyt @Abhinandan-Purkait ?

File: tests/bdd/features/volume/topology/pool-affinity.feature
Comment (tiagolobocastro) [inline tests/bdd/features/volume/topology/pool-affinity.feature:1]: This is called affinity but also seems to include exclusion, is this intended?

Comment (sinhaashish) [reply tests/bdd/features/volume/topology/pool-affinity.feature:1]: renamed it to `pool-affinity-spread.feature` and `test_pool_affinity_spread.py`

File: tests/bdd/features/volume/topology/pool-affinity.feature (hunk 24 above)
Comment (sinhaashish) [reply tests/bdd/features/volume/topology/pool-affinity.feature]: IMHO, there are 4 test cases only , This PR can be taken as is . Based on time i can try out the  Scenario Outline and Examples

File: tests/bdd/features/volume/topology/pool-affinity.feature (hunk 23 above)
Comment (sinhaashish) [reply tests/bdd/features/volume/topology/pool-affinity.feature]: Made the changes, Thanks for pointing out. Updated the test case as will

File: control-plane/agents/src/bin/core/controller/scheduling/volume_policy/pool.rs (hunk 18 above)
Comment (sinhaashish) [reply control-plane/agents/src/bin/core/controller/scheduling/volume_policy/pool.rs]: removed all these logic and replace with `does_pool_qualify_inclusion_labels` and `does_pool_qualify_exclusion_labels` functions

File: control-plane/agents/src/bin/core/controller/scheduling/volume_policy/pool.rs (hunk 17 above)
Comment (sinhaashish) [reply control-plane/agents/src/bin/core/controller/scheduling/volume_policy/pool.rs:145]: removes all these logic and replace with does_pool_qualify_inclusion_labels and does_pool_qualify_exclusion_labels functions

File: control-plane/agents/src/bin/core/controller/scheduling/volume_policy/pool.rs (hunk 14 above)
Comment (sinhaashish) [reply control-plane/agents/src/bin/core/controller/scheduling/volume_policy/pool.rs]: Thanks for pointing out . Modified it

File: control-plane/agents/src/bin/core/controller/scheduling/volume_policy/pool.rs [hunk 25]
```diff
@@ -73,41 +73,105 @@ impl PoolBaseFilters {
...
     }
 }
+
+/// Return true if all the key present in volume's pool inclusion matches with the pool labels
+/// otherwise returns false
+pub(crate) fn does_pool_qualify_inclusion_labels(
+    vol_pool_inc_labels: HashMap<String, String>,
+    pool_labels: HashMap<String, String>,
+) -> bool {
+    let mut inc_match = true; // Initialize to true, assuming inclusive match until proven otherwise
```
Comment (tiagolobocastro) [inline control-plane/agents/src/bin/core/controller/scheduling/volume_policy/pool.rs]: Don't need this, can just return false directly instead of breaking?

File: control-plane/agents/src/bin/core/controller/scheduling/volume_policy/pool.rs [hunk 26]
```diff
@@ -73,41 +73,105 @@ impl PoolBaseFilters {
...
+    inc_match
+}
+
+/// Return true if all the key present in volume's pool exclusion matches with the pool labels and
+/// their corresponding value are different otherwise returns false
+pub(crate) fn does_pool_qualify_exclusion_labels(
+    vol_pool_exc_labels: HashMap<String, String>,
+    pool_labels: HashMap<String, String>,
+) -> bool {
+    // If
```
Comment (tiagolobocastro) [inline control-plane/agents/src/bin/core/controller/scheduling/volume_policy/pool.rs]: Missing rest of comment? :)

**Review by tiagolobocastro — COMMENTED** [review]
I don't think this is being catered for right?

> nodeTopologyAffinity: "rack"
replica1: "rack: A"    <- A determined at creation time
replica2: "rack: A"

**Comment by sinhaashish** [comment]
> I don't think this is being catered for right?
> 
> > nodeTopologyAffinity: "rack"
> > replica1: "rack: A"    <- A determined at creation time
> > replica2: "rack: A"

This PR just caters to below scenario
poolTopologyAffinity: "performance: Gold"
replica1: "performance: Gold"
replica2: "performance: Gold"

poolTopologySpread: "storage: Hot". 
replica1: "storage: Cold"
replica2: "storage: Warm"


This is pool topology , not node .

**Comment by tiagolobocastro** [comment]
Is that not the same for pools?

> poolTopologyAffinity: "foo"
replica1: "foo: doo" <- A determined at creation time
replica2: "foo: bah"

File: control-plane/agents/src/bin/core/controller/scheduling/volume_policy/pool.rs [hunk 27]
```diff
@@ -73,41 +73,127 @@ impl PoolBaseFilters {
...
+                    inc_match && exc_match
+                }
             },
             Err(_) => false,
         }
     }
 }
+
+/// Return true if all the key present in volume's pool inclusion matches with the pool labels
+/// otherwise returns false
```
Comment (tiagolobocastro) [inline control-plane/agents/src/bin/core/controller/scheduling/volume_policy/pool.rs]:
```suggestion
/// Return true if all the keys present in volume's pool inclusion matches with the pool labels
/// otherwise return false.
```

File: control-plane/agents/src/bin/core/controller/scheduling/volume_policy/pool.rs [hunk 28]
```diff
@@ -73,41 +73,127 @@ impl PoolBaseFilters {
...
+        // If volume's pool exclusion labels has any key with empty value, then it means that the
+        // pool should not have that key
+        for key in excluion_key_only.iter() {
+            if pool_labels.contains_key(&key.to_string()) {
+                return false;
+            }
+        }
+    }
+
+    true
```
Comment (tiagolobocastro) [inline control-plane/agents/src/bin/core/controller/scheduling/volume_policy/pool.rs]:
```suggestion
        let common_keys: Vec<_> = vol_pool_exc_labels
            .keys()
            .filter(|&key| pool_labels.contains_key(key))
            .collect();

        if common_keys.is_empty() {
            return false;
        }

        common_keys
            .iter()
            .any(|&key| vol_pool_exc_labels.get(key) != pool_labels.get(key))
    } else {
        // If volume's pool exclusion labels has any key with empty value, then it means that the
        // pool should not have that key
        !exclusion_key_only.iter().any(|key| pool_labels.contains_key(&key.to_string()))
    }
```

File: control-plane/agents/src/bin/core/controller/scheduling/volume_policy/pool.rs [hunk 29]
```diff
@@ -73,41 +73,127 @@ impl PoolBaseFilters {
...
+
+    // excluion_key_only are set of keys from vol_pool_exc_labels has only keys with empty values
+    // e.g. vol_pool_exc_labels = {"key1": "", "key2": "value2"}
+    let excluion_key_only: Vec<_> = vol_pool_exc_labels
+        .iter()
+        .filter(|(_, value)| value.is_empty())
+        .map(|(key, _)| key)
+        .collect();
+
+    if excluion_key_only.is_empty() {
```
Comment (tiagolobocastro) [inline control-plane/agents/src/bin/core/controller/scheduling/volume_policy/pool.rs]:
```suggestion
    let exclusion_key_only: Vec<_> = vol_pool_exc_labels
        .iter()
        .filter(|(_, value)| value.is_empty())
        .map(|(key, _)| key)
        .collect();

    if exclusion_key_only.is_empty() {
```

File: control-plane/agents/src/bin/core/controller/scheduling/volume_policy/pool.rs [hunk 30]
```diff
@@ -73,41 +73,127 @@ impl PoolBaseFilters {
...
+pub(crate) fn does_pool_qualify_exclusion_labels(
+    vol_pool_exc_labels: HashMap<String, String>,
+    pool_labels: HashMap<String, String>,
+) -> bool {
+    // If volume's pool exclusion labels is empty, then it means that the volume has no exclusion
+    if vol_pool_exc_labels.is_empty() {
+        return true;
+    }
+
+    // excluion_key_only are set of keys from vol_pool_exc_labels has only keys with empty values
```
Comment (tiagolobocastro) [inline control-plane/agents/src/bin/core/controller/scheduling/volume_policy/pool.rs]:
```suggestion
    // exclusion_key_only are set of keys from vol_pool_exc_labels has only keys with empty values
```

File: tests/bdd/features/volume/topology/test_pool_affinity_spread.py [hunk 31]
```diff
@@ -0,0 +1,642 @@
...
+        POOL_4_UUID: ApiClient.pools_api().get_pool(POOL_4_UUID)["spec"]["labels"],
+    }
+    qualified_pools = list()
+    for pool_id, labels in pool_with_labels.items():
+        if does_pool_qualify_inclusion_labels(volume_pool_topology_inclusion_labels, labels ) and does_pool_qualify_exclusion_labels(volume_pool_topology_exclusion_labels, labels ):
+            qualified_pools.append(pool_id)
+    return  len(qualified_pools) 
+
+
+def does_pool_qualify_inclusion_labels(volume_pool_topology_inclusion_labels, pool_labels):
```
Comment (tiagolobocastro) [inline tests/bdd/features/volume/topology/test_pool_affinity_spread.py:611]: I still think it'd be better to explicitly compare with actual examples:
1. it'll be a lot clearer to see what is being tested exactly, with examples
2. less reliant on test code to ensure the functionality is correct

It can be done on another PR but this should be next PR IMHO

File: control-plane/csi-driver/src/context.rs [hunk 32]
```diff
@@ -116,6 +143,18 @@ impl Parameters {
     pub fn fs_id(value: Option<&String>) -> Result<Option<Uuid>, UuidError> {
         Self::parse_uuid(value)
     }
+    /// Parse the value for `Self::PoolTopologyAffinity`.
```
Comment (tiagolobocastro) [inline control-plane/csi-driver/src/context.rs:152]: Please doc comment here various examples of value

**Comment by sinhaashish** [comment]
> Is that not the same for pools?
> 
> > poolTopologyAffinity: "foo"
> > replica1: "foo: doo" <- A determined at creation time
> > replica2: "foo: bah"

As per the initial design it was not , but to make poolTopologyAffinity consistent with nodeTopologyAffinity
I have added the same

**Review by tiagolobocastro — APPROVED** [review]

**Review by Abhinandan-Purkait — APPROVED** [review]

**Comment by sinhaashish** [comment]
bors merge

**Comment by sinhaashish** [comment]
bors merge

**Comment by tiagolobocastro** [comment]
@sinhaashish for the next PR please do not address comments in a separate commit, rather squash them into where the code applies
Also refactoring is its own subject, so rather than
feat(topology): modularize the pool operator 
it should be
refactor(topology): modularize the pool operator
